improver/utilities/neighbourhood_tools.py

```diff
@@ -0,0 +1,106 @@
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Provides tools for neighbourhood generation"""
+
+import numpy as np
+
+
+def rolling_window(input_array, shape, writeable=False):
+    """Creates a rolling window neighbourhoods of the given `shape` from the
+    last `len(shape)` axes of the input array. avoids creating large output
+    array by constructing a non-continuous view mapped onto the input array.
+
+    args:
+        input_array (numpy.ndarray):
+            A 2-D array padded with nans for half the
+            neighbourhood size.
+        shape (tuple(int)):
+            The neighbourhood shape e.g. is the neighbourhood
+            size is 3, the shape would be (3, 3) to create a
+            3x3 array around each point in the input_array.
+        writeable (bool):
+            If True the returned view will be writeable. This will modify
+            the input array, so use with caution.
+
+    Returns:
+        numpy.ndarray:
+            "views" into the data, each view represents
+            a neighbourhood of points.
+    """
+    num_window_dims = len(shape)
+    num_arr_dims = len(input_array.shape)
+    assert num_arr_dims >= num_window_dims
+    adjshp = (
+        *input_array.shape[:-num_window_dims],
+        *(
+            arr_dims - win_dims + 1
+            for arr_dims, win_dims in zip(input_array.shape[-num_window_dims:], shape)
+        ),
+        *shape,
+    )
+    assert all(arr_dims > 0 for arr_dims in adjshp)
+    strides = input_array.strides + input_array.strides[-num_window_dims:]
+    return np.lib.stride_tricks.as_strided(
+        input_array, shape=adjshp, strides=strides, writeable=writeable
+    )
+
+
+def pad_and_roll(input_array, shape, **kwargs):
+    """Pads the last `len(shape)` axes of the input array for `rolling_window`
+    to create 'neighbourhood' views of the data of a given `shape` as the last
+    axes in the returned array. Collapsing over the last `len(shape)` axes
+    results in a shape of the original input array.
+
+    args:
+        input_array (numpy.ndarray):
+            The dataset of points to pad and create rolling windows for.
+        shape (tuple(int)):
+            Desired shape of the neighbourhood. E.g. if a neighbourhood
+            width of 1 around the point is desired, this shape should be (3, 3):
+                  X X X
+                  X O X
+                  X X X
```

Comment:
Sphinx is picky about these kinds of things. You need double colons to indicate something like a code snippet (which will keep the formatting as you expect) and a clear line above and below the snippet. Without the change it looks like this in the docs:

![unfixed](https://user-images.githubusercontent.com/28042362/80485550-8aaa6300-8951-11ea-8886-0f132dea4570.PNG)

With the change it looks like this:

![fixed](https://user-images.githubusercontent.com/28042362/80485593-9bf36f80-8951-11ea-9e79-bb269fd3a1d5.PNG)


```suggestion
            width of 1 around the point is desired, this shape should be (3, 3)::
            
                  X X X
                  X O X
                  X X X
                  
```